sdk/storage/Azure.Storage.Files.Shares/src/Models/ShareAccessPolicy.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Text;
+
+namespace Azure.Storage.Files.Shares.Models
+{
+    /// <summary>
+    /// Share Access policy.
+    /// </summary>
+    public partial class ShareAccessPolicy
+    {
+        /// <summary>
+        /// The date-time the policy is active.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public System.DateTimeOffset StartsOn
+        {
+            get
+            {
+                return PolicyStartsOn == default ?
+                    StartsOn :
+                    (DateTimeOffset)PolicyStartsOn;
+            }
+            set
+            {
+                PolicyStartsOn = value;
+            }
+        }
+
+        /// <summary>
+        /// The date-time the policy expires.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public System.DateTimeOffset ExpiresOn
```

Comment:
In the currently released SDK, if a user doesn't set the `ExpiresOn` property, we will still send default(DateTimeOffset) to the server, right? With the change here, we would stop sending anything so it would break their scenarios, if they were relying on this data getting set.
This was the motivation behind introducing a constructor with the properties as parameters. If the constructor is used, we can give the nice behavior of omitting unset properties. If it isn't used, we would retain the old behavior.

Reply:
got it

Reply:
I think the old behavior was a bug, default(DateTimeOffset) is not a valid value for ExpiresOn, and probably not a what the user intended for StartsOn.

Reply:
If you set to default, does the service throw an exception?

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/13058

Reply:
>Start and expiration is not nullable for ShareAccessPolicy, if no value assigned to them, default DateTimeOffset value will be used.

Right, so the service would happily accept the default DateTimeOffset as the starts or expiry time if a user leaves the value unset. Even if this behavior is a bug, customers might be relying on it now.

Reply:
I really don't think users intend to set `ExpiresOn` and `StartsOn` to `1/1/0001 12:00:00 AM +00:00` when they don't set either value.  Especially for `ExpiresOn`, the current behavior results in an access policy that doesn't work by default.

Reply:
I don't think we should be defining the `ExpiresOn` or `StartsOn` for the customer for them if they don't explicitly state it. In the v11 SDK, and if the user were to define their own Access Policy, they should expect that they're optional parameters.

Reply:
Yep, I think the whole reason for this change is to make `ExpiresOn` and `StartsOn` optional.

Reply:
Yes, totally agree that we want them to be optional. The question is whether we are willing to change existing behavior for customers that are not using the new properties/ctor. The main scenario I am worried about is customers are relying on this behavior to rescind access by setting `ExpiresOn` to default.

Reply:
As folks mentioned, what we send today by default doesn't really make sense. So customers relying on that are most likely have a bug in their code.

Reply:
Let's say I'm a customer who doesn't want to set either `ExpiresOn` or `StartsOn`.  How would I specify this behavior if we are setting both to DateTimeOffset.Min if they aren't provided?

I think it's unlikely users have taken a dependency on our current (broken) behavior.

Reply:
Here is the scenario that could work:
- when creating an AccessPolicy the customer sends both ExpiresOn and StartsOn
- when rescinding the AccessPolicy, they just leave ExpiresOn and StartsOn blank since this will reset to default and make the policy no longer active.

Reply:
https://github.com/Azure/azure-sdk-for-net/issues/13058

> For SetAccessPolicyAsync of ShareClient, there is no way to not set expiration time of a single ShareAccessPolicy.

How would the customer be able to accomplish this if we set `ExpiresOn` to `DateTimeOffset.Min` if it is not provided?

Reply:
We wouldn't set it to the default if they use the new constructor.

Reply:
Is permissions required? I would say we can just have that as the sole param in the ctor.

Reply:
Permissions is not required. All three are optional

Reply:
Ah, in that case, I don't see a better option than listing all three as optional params.

Reply:
Is there a way to somehow take out the default constructor `public ShareAccessPolicy()` from the generated code?@JoshLove-msft . I'm trying to make sure that the old behavior where if they use the default constructor, that the value of `default(DateTimeOffset.Min)` gets set in the new `PolicyStartsOn` and `PolicyExpiresOn` and `StartsOn` and `ExpiresOn`. Also I want to be able to mark it with the `EditorBrowsable(Never)`

Reply:
I'm not sure if this is possible. I think we would have to create a new internal type and just manually add the default constructor in a hand-written model. This means we would have to do the same for SignedIdentifiers as well. Before we go down this road, would you be able to write up a test that either validates or invalidates the scenario that I proposed where this could break existing users? Could a user rescind access for an already created identifier by passing an AccessPolicy that leaves the StartsOn/ExpiresOn dates null? If the ExpiresOn date is set to default, will this prevent a user from using the SAS?

Reply:
I'm not sure if this is possible. I think we would have to create a new internal type and just manually add the default constructor in a hand-written model. This means we would have to do the same for SignedIdentifiers as well. Before we go down this road, would you be able to write up a test that either validates or invalidates the scenario that I proposed where this could break existing users? Could a user rescind access for an already created identifier by passing an AccessPolicy that leaves the StartsOn/ExpiresOn dates null? If the ExpiresOn date is set to default, will this prevent a user from using the SAS?

Reply:
So I tested this out and if you don't have the `ExpiresOn` date in the SAS, then it is required in the access policy. This means there should be no practical difference between sending default, and omitting the `ExpiresOn` from the policy. `StartsOn` is not required, but the default and null are effectively the same. This means that if we want to avoid making any API change, we have that option if the generator can be updated to not send default values (as Ted originally suggested). I think having these properties be nullable is still probably more correct, but now we can weigh this against whether it is worth introducing new properties here. 

```c#
await using DisposingContainer test = await GetTestContainerAsync();
var id = "myCoolId";
BlobSignedIdentifier identifier = new BlobSignedIdentifier
{
    AccessPolicy = new BlobAccessPolicy()
    {
        ExpiresOn = DateTime.UtcNow.AddDays(1)
    },
    Id = id
};
await test.Container.SetAccessPolicyAsync(permissions: new BlobSignedIdentifier[] { identifier });
var builder = new BlobSasBuilder
{
    BlobContainerName = test.Container.Name,
    Identifier = id
};
builder.SetPermissions(BlobContainerSasPermissions.List);
var uri = new BlobUriBuilder(test.Container.Uri)
{
    Sas = builder.ToSasQueryParameters(GetNewSharedKeyCredentials())
};
var sasClient = new BlobContainerClient(uri.ToUri());

// this works as the expiration is in the future and we have the necessary permissions
var blobs = await sasClient.GetBlobsAsync().ToListAsync();

// revoke access
var revoke = new BlobSignedIdentifier
{
    // send an empty access policy with the same Id to revoke access
    AccessPolicy = new BlobAccessPolicy(),
    Id = id
};
await test.Container.SetAccessPolicyAsync(permissions: new BlobSignedIdentifier[] { revoke });
// seems to be a lag before the new access policy takes effect
await Task.Delay(TimeSpan.FromSeconds(30));
// this throws as access has been revoked regardless of whether we send 
// default ExpiresOn or omit it entirely
blobs = await sasClient.GetBlobsAsync().ToListAsync();
```

Reply:
I think probably doing the generator enhancement would be nice, but I think it might be faster to do the internal copy of `ShareAccessPolicy` (and `ShareSignedIndentifiers`). I think from a user perspective, it would be better to not have API changes, because we could run into the question and confusing scenario of a customer setting both the `StartsOn` and `PolicyStartsOn` even though one is marked with `EditorBrowserable(Never)`. This would be my preference.

Is the advantage to having the API change is that the user is able to set the `PolicyStartsOn` and `PolicyExpiresOn` to `null`, instead of us evaluating `StartsOn` and `ExpiresOn` value of `default(DateTimeOffset)` aka `DateTimeOffset.MinValue` as the default value and assuming that if the value was set to that, that would mean the user means `null` ?

@JoshLove-msft Do you guys have a preference on what route we should go?

Reply:
Yeah, the main advantage is that we can actually represent the true state of what is on the server with nullable properties.

Reply:
I just talked with Ted and we both think that the way you have it in the PR now makes the most sense.

Reply:
Okay I will go ahead with the changes to Blob, DataLake and Queues. Thanks!